FeedStoreChallenge/CoreDataFeedStore.swift

```diff
@@ -29,14 +29,52 @@ public final class CoreDataFeedStore: FeedStore {
 	}
 
 	public func retrieve(completion: @escaping RetrievalCompletion) {
-		fatalError("Must be implemented")
+		perform { context in
+			do {
+				if let result = try ManagedCache.cache(in: context) {
+					completion(.found(feed: result.localImages, timestamp: result.timestamp))
+				} else {
+					completion(.empty)
+				}
+			} catch {
+				completion(.failure(error))
+			}
+		}
 	}
 
 	public func insert(_ feed: [LocalFeedImage], timestamp: Date, completion: @escaping InsertionCompletion) {
-		fatalError("Must be implemented")
+		perform { context in
+			do {
+				let managedCache = try ManagedCache.new(in: context)
+				managedCache.timestamp = timestamp
+				managedCache.feed = ManagedFeedImage.imageSet(from: feed, in: context)
+
+				try context.save()
+				completion(nil)
+			} catch {
+				context.rollback()
+				completion(error)
+			}
+		}
 	}
 
 	public func deleteCachedFeed(completion: @escaping DeletionCompletion) {
-		fatalError("Must be implemented")
+		perform { context in
+			do {
+				try ManagedCache.delete(in: context)
+				try context.save()
```

Comment:
In case the cache is empty, `ManagedCache.delete` will not delete anything or change the managed context. Thus, there will be no need to call `save`. 

Core Data may perform unnecessary work when calling save, even if there are no changes.

You can check the `NSManagedObjectContext.hasChanges`property before calling save to avoid unnecessary work:

```suggestion
				try ManagedCache.delete(in: context)
				if context.hasChanges { 
					try context.save()
				}
```